server/handlers.go

```diff
@@ -319,12 +330,18 @@ func (s *Server) handleConnectorLogin(w http.ResponseWriter, r *http.Request) {
 func (s *Server) handleConnectorCallback(w http.ResponseWriter, r *http.Request) {
 	var authID string
 	switch r.Method {
-	case "GET": // OAuth2 callback
-		if authID = r.URL.Query().Get("state"); authID == "" {
+	case http.MethodGet: // OAuth2 or SAML HTTP-Redirect callback
+		// try OAuth2
+		authID = r.URL.Query().Get("state")
```

Comment:
The SAML connector can still use the value set by "state" here. We need to enforce:

```
var usedRelayState bool
authID = r.URL.Query().Get("state")
if authID == "" {
    usedRelayState = true
    authID = r.URL.Query.Get("RelayState")
}
```  

Then verify this value later.

```
case connector.CallbackConnector:
    if usedRelayState {
        return nil, fmt.Errorf("no state parameter provided")
    }
case connector.SAMLConnector:
    if !userRelayState {
        return nil, fmt.Errorf("invalid url query value 'state' provided")
    }

```

Reply:
understand!